pytorch_lightning/callbacks/swa.py

```diff
@@ -0,0 +1,234 @@
+                return True
+        return False
+
+    # Inspiration from
+    def reset_batch_norm_and_save_state(self, average_model, device):
+        """
+        Credit to PyTorch Team.
+        Adapted from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L115
+        """
+        self.momenta = {}
+        for module in average_model.modules():
+            if isinstance(module, nn.modules.batchnorm._BatchNorm):
+                running_mean_dtype = module.running_mean.dtype
+                running_var_dype = module.running_var.dtype
+                module.running_mean = torch.zeros_like(module.running_mean, device=device, dtype=running_mean_dtype)
+                module.running_var = torch.ones_like(module.running_var, device=device, dtype=running_var_dype)
+                self.momenta[module] = module.momentum
+                module.momentum = None
+                module.num_batches_tracked *= 0
+
+    def on_fit_start(self, trainer, pl_module):
+        self._average_model = deepcopy(pl_module).to("cpu")
+        optimizers = trainer.optimizers
+        lr_schedulers = trainer.lr_schedulers
+
+        if len(optimizers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 optimizer.")
+
+        if len(lr_schedulers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 lr_scheduler.")
+
+        self._max_epochs = trainer.max_epochs
+
+        # convert float to integer.
+        if isinstance(self._swa_epoch_start, float):
+            self._swa_epoch_start = int(self._max_epochs * self._swa_epoch_start)
+
+        self._model_contains_batch_norm = self.pl_module_contains_batch_norm(pl_module)
+
+        if self._model_contains_batch_norm:
+            # virtually increase max_epochs to perform batch norm update on latest epoch.
+            trainer.max_epochs += 1
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        if trainer.current_epoch == self._swa_epoch_start:
+            optimizers = trainer.optimizers
+            lr_scheduler = trainer.lr_schedulers[0]["scheduler"]
+
+            self._swa_scheduler = SWALR(
+                optimizers[0],
+                swa_lr=self._swa_lrs,
+                anneal_epochs=self._annealing_epochs,
+                anneal_strategy=self._annealing_strategy,
+                last_epoch=trainer.max_epochs if self._annealing_strategy == "cos" else -1
+            )
+
+            rank_zero_warn(f"swapping lr_scheduler {lr_scheduler} for {self._swa_scheduler}")
+
+            trainer.lr_schedulers[0]["scheduler"] = self._swa_scheduler
+
+            self.n_averaged = torch.tensor(0, dtype=torch.long, device=pl_module.device)
+
+        elif self._model_contains_batch_norm and trainer.current_epoch == self._max_epochs:
+            # Turn off backwards and optimization, only update batch norm stats
+            self.transfer_weights(self._average_model, pl_module)
+
```

Comment:
So why do we need `transfer_weights`? After this `elif` is run, the `if` right after and `update_parameters` will copy `p_model` weights to `p_swa`

Reply:
Good catch. There were a bug.